Automattic/wp-calypso
# State: Update counts state in response to post status changes

Related: #5613

This pull request seeks to enhance the `state.posts.counts` state to account for changes in post status, transitioning counts from its previous status to the new status to reflect the change.

__Implementation notes:__

This is a difficult challenge to solve, since post counts are received and tracked in state independently of the posts themselves. This is desirable since when navigating to a post listing screen, the filter tabs should display the counts of each status, even before all posts of that status are requested. This poses a problem when posts change over time, since there's effectively a dependency between the two states. When deleting a post or receiving a new post, the previous implementation had no way of determining what the post's previous status was, since that was tracked elsewhere in state.

To address this, I've experimented with the idea of having the reducer track its own local state, monitoring actions that it considers relevant to count changes. Here, this includes trashing posts (`POST_DELETE` action) and changes to posts over time (`POSTS_RECEIVE`). Further, since counts are distinguished by "mine" and "all", the reducer must also track the current user ID as it changes.

While the implementation is quite different, this was very loosely inspired by an `undoable` higher-order reducer demo'd during this year's React Europe conference ([video](https://www.youtube.com/watch?v=uvAXVMwHJXU&feature=youtu.be&list=PLCC436JpVnK09bZeayg-KeLuHfHgc-tDa&t=677)).

__Testing instructions:__

Ensure that Mocha tests pass:

```
npm run test-client client/state/posts/counts/test/reducer.js
```

Observe that when trashing a post from the [CPT listing screen](http://calypso.localhost:3000/types/post), the post counts in both the filter bar and sidebar draft button update accordingly. Specifically, note that posts and pages support "trash", while other post types do not, so "trash" count is only incremented for these types. Additionally, when trashing an already-trashed post, the post count for trashed should decrement.

You can also verify that publishing a draft from the editor updates these counts.

__Caveats:__

- Not all posts are tracked in Redux state. Notably, the (non-CPT) post listing screen still uses legacy Flux stores, so the draft count in the sidebar will not update when trashing a post from this screen

/cc @mtias 

Test live: https://calypso.live/?branch=update/cpt-increment-post-counts
